File: router.py
#!/usr/bin/env python3
"""Shortest path and novelty-weighted pedestrian routing.

Uses parent-pointer A* with int32 array indices for O(V) memory
instead of O(V^2) path copies in every queue entry.
"""

import heapq
import math

import numpy as np

from graph_builder import haversine, bearing


def _heuristic_idx(G, idx, target_lat, target_lon):
    """A* heuristic: straight-line distance from node index to target coords."""
    return haversine(float(G.node_lats[idx]), float(G.node_lons[idx]),
                     target_lat, target_lon)


def _reconstruct_path(came_from, src_idx, tgt_idx):
    """Walk parent pointers from target back to source. Returns list of indices."""
    path = []
    cur = tgt_idx
    while cur != -1:
        path.append(cur)
        if cur == src_idx:
            break
        cur = int(came_from[cur])
    path.reverse()
    return path


def _path_distance(G, path_indices):
    """Sum edge weights along a path of node indices."""
    total = 0.0
    for i in range(len(path_indices) - 1):
        u = path_indices[i]
        targets, weights = G.neighbors(u)
        v = path_indices[i + 1]
        # Find the edge to v
        mask = targets == v
        total += float(weights[mask][0])
    return total


def shortest_path(G, source_osm, target_osm):
    """Find shortest path using A* with haversine heuristic.

    Args:
        G: CompactGraph
        source_osm: Source node OSM ID
        target_osm: Target node OSM ID

    Returns:
        (path, distance) where path is a list of OSM node IDs and distance is in meters.
        Returns (None, None) if no path exists.
    """
    if source_osm == target_osm:
        return [source_osm], 0.0

    src_idx = G.idx_for_osm_id(source_osm)
    tgt_idx = G.idx_for_osm_id(target_osm)

    target_lat = float(G.node_lats[tgt_idx])
    target_lon = float(G.node_lons[tgt_idx])
    num_nodes = G.number_of_nodes()

    came_from = np.full(num_nodes, -1, dtype=np.int32)
    g_score = np.full(num_nodes, np.inf, dtype=np.float64)
    g_score[src_idx] = 0.0

    counter = 0
    # Priority queue: (f_score, g_score, counter, node_idx)
    open_set = [(0.0, 0.0, counter, src_idx)]

    while open_set:
        f, g, _, current = heapq.heappop(open_set)

        if current == tgt_idx:
            path_indices = _reconstruct_path(came_from, src_idx, tgt_idx)
            path_osm = [int(G.node_ids[i]) for i in path_indices]
            return path_osm, g

        if g > g_score[current]:
            continue

        targets, weights = G.neighbors(current)
        for j in range(len(targets)):
            neighbor = int(targets[j])
            new_g = g + float(weights[j])

            if new_g < g_score[neighbor]:
                g_score[neighbor] = new_g
                came_from[neighbor] = current
                h = _heuristic_idx(G, neighbor, target_lat, target_lon)
                counter += 1
                heapq.heappush(open_set, (new_g + h, new_g, counter, neighbor))

    return None, None


def path_to_edges(path):
    """Convert a node path to a list of edge tuples (OSM IDs)."""
    return [(path[i], path[i + 1]) for i in range(len(path) - 1)]


def _edge_key(n1, n2):
    """Canonical edge key for undirected lookup."""
    return (min(n1, n2), max(n1, n2))


def novelty_route(G, source, target, walked_edges, min_novelty=0.3, max_overhead=0.25):
    """Find a route that maximizes novel (unwalked) edges.

    Args:
        G: CompactGraph
        source: Start node OSM ID
        target: End node OSM ID
        walked_edges: Set of (min_node, max_node) edge keys that have been walked
        min_novelty: Minimum fraction of edges that should be novel (0.0 - 1.0)
        max_overhead: Maximum allowed overhead vs shortest path (0.0 - 1.0)

    Returns:
        dict with keys: path, distance, novelty, overhead, shortest_distance, edges
    """
    # Phase 1: Get baseline shortest path
    base_path, base_dist = shortest_path(G, source, target)
    if base_path is None:
        return None

    base_edges = path_to_edges(base_path)
    base_novel = _compute_novelty(base_edges, walked_edges)

    if base_novel >= min_novelty:
        return _build_result(base_path, base_dist, base_dist, walked_edges, G)

    if not walked_edges:
        return _build_result(base_path, base_dist, base_dist, walked_edges, G)

    # Phase 2: Iterative penalty search
    best_result = None
    best_novelty = base_novel

    lo_penalty = 1.0
    hi_penalty = 10.0

    for _ in range(5):
        path, dist = _penalized_astar(G, source, target, walked_edges, hi_penalty)
        if path is None:
            hi_penalty = (lo_penalty + hi_penalty) / 2
            continue
        edges = path_to_edges(path)
        novelty = _compute_novelty(edges, walked_edges)
        if novelty >= min_novelty:
            break
        hi_penalty *= 2
        if hi_penalty > 100:
            break

    for _ in range(10):
        mid_penalty = (lo_penalty + hi_penalty) / 2
        path, dist = _penalized_astar(G, source, target, walked_edges, mid_penalty)

        if path is None:
            hi_penalty = mid_penalty
            continue

        edges = path_to_edges(path)
        novelty = _compute_novelty(edges, walked_edges)
        overhead = (dist - base_dist) / base_dist if base_dist > 0 else 0

        if overhead <= max_overhead and novelty > best_novelty:
            best_novelty = novelty
            best_result = _build_result(path, dist, base_dist, walked_edges, G)

        if novelty < min_novelty:
            lo_penalty = mid_penalty
        elif overhead > max_overhead:
            hi_penalty = mid_penalty
        else:
            lo_penalty = mid_penalty

    if best_result is None or best_result["novelty"] < min_novelty:
        for penalty in [1.5, 2.0, 3.0, 5.0, 8.0]:
            path, dist = _penalized_astar(G, source, target, walked_edges, penalty)
            if path is None:
                continue
            edges = path_to_edges(path)
            novelty = _compute_novelty(edges, walked_edges)
            overhead = (dist - base_dist) / base_dist if base_dist > 0 else 0

            if overhead <= max_overhead and novelty > best_novelty:
                best_novelty = novelty
                best_result = _build_result(path, dist, base_dist, walked_edges, G)

    if best_result is None:
        best_result = _build_result(base_path, base_dist, base_dist, walked_edges, G)

    return best_result


def _penalized_astar(G, source_osm, target_osm, walked_edges, penalty_factor):
    """A* search with penalty on walked edges. Uses parent pointers."""
    if source_osm == target_osm:
        return [source_osm], 0.0

    src_idx = G.idx_for_osm_id(source_osm)
    tgt_idx = G.idx_for_osm_id(target_osm)

    target_lat = float(G.node_lats[tgt_idx])
    target_lon = float(G.node_lons[tgt_idx])
    num_nodes = G.number_of_nodes()

    came_from = np.full(num_nodes, -1, dtype=np.int32)
    g_score = np.full(num_nodes, np.inf, dtype=np.float64)
    g_score[src_idx] = 0.0

    counter = 0
    open_set = [(0.0, 0.0, counter, src_idx)]

    while open_set:
        f, g, _, current = heapq.heappop(open_set)

        if current == tgt_idx:
            # Reconstruct path and compute actual (unpenalized) distance
            path_indices = _reconstruct_path(came_from, src_idx, tgt_idx)
            path_osm = [int(G.node_ids[i]) for i in path_indices]
            actual_dist = _path_distance(G, path_indices)
            return path_osm, actual_dist

        if g > g_score[current]:
            continue

        current_osm = int(G.node_ids[current])
        targets, weights = G.neighbors(current)
        for j in range(len(targets)):
            neighbor = int(targets[j])
            edge_weight = float(weights[j])

            # Apply penalty to walked edges (using OSM IDs)
            neighbor_osm = int(G.node_ids[neighbor])
            ek = _edge_key(current_osm, neighbor_osm)
            effective_weight = edge_weight
            if ek in walked_edges:
                effective_weight *= penalty_factor

            new_g = g + effective_weight

            if new_g < g_score[neighbor]:
                g_score[neighbor] = new_g
                came_from[neighbor] = current
                h = _heuristic_idx(G, neighbor, target_lat, target_lon)
                counter += 1
                heapq.heappush(open_set, (new_g + h, new_g, counter, neighbor))

    return None, None


def _compute_novelty(edges, walked_edges):
    """Compute fraction of edges that are novel (not walked)."""
    if not edges:
        return 1.0
    novel = sum(1 for e in edges if _edge_key(*e) not in walked_edges)
    return novel / len(edges)


def _build_result(path, distance, base_distance, walked_edges, G=None):
    """Build a route result dictionary."""
    edges = path_to_edges(path)
    novelty = _compute_novelty(edges, walked_edges)
    overhead = (distance - base_distance) / base_distance if base_distance > 0 else 0

    result = {
        "path": path,
        "edges": edges,
        "distance": distance,
        "shortest_distance": base_distance,
        "novelty": novelty,
        "overhead": overhead,
    }

    if G is not None and G.has_name_data():
        result["instructions"] = generate_instructions(path, G)

    return result


# Highway type fallback descriptions
_HIGHWAY_DESCRIPTIONS = {
    "footway": "footpath",
    "path": "path",
    "pedestrian": "pedestrian way",
    "steps": "steps",
    "cycleway": "cycleway",
    "residential": "road",
    "living_street": "road",
    "tertiary": "road",
    "tertiary_link": "road",
    "secondary": "road",
    "secondary_link": "road",
    "primary": "road",
    "primary_link": "road",
    "trunk": "road",
    "service": "service road",
    "track": "track",
    "unclassified": "road",
}


def _compass_direction(bearing_deg):
    """Convert bearing in degrees to compass direction string."""
    dirs = ["north", "northeast", "east", "southeast",
            "south", "southwest", "west", "northwest"]
    idx = int((bearing_deg + 22.5) % 360 / 45) % 8
    return dirs[idx]


def _classify_turn(angle):
    """Classify turn from signed angle (negative=left, positive=right)."""
    abs_angle = abs(angle)
    if abs_angle < 15:
        return "straight"
    elif abs_angle < 45:
        return "slight_left" if angle < 0 else "slight_right"
    elif abs_angle < 120:
        return "left" if angle < 0 else "right"
    elif abs_angle < 160:
        return "sharp_left" if angle < 0 else "sharp_right"
    else:
        return "u_turn"


_TURN_PREFIXES = {
    "start": "Head",
    "straight": "Continue",
    "slight_left": "Turn slight left",
    "slight_right": "Turn slight right",
    "left": "Turn left",
    "right": "Turn right",
    "sharp_left": "Turn sharp left",
    "sharp_right": "Turn sharp right",
    "u_turn": "Make a U-turn",
    "arrive": "Arrive",
}


def generate_instructions(path, G):
    """Generate turn-by-turn navigation instructions from a route path.

    Args:
        path: List of OSM node IDs
        G: CompactGraph with name data

    Returns:
        List of instruction dicts, or None if no name data
    """
    if not G.has_name_data():
        return None
    if len(path) < 2:
        return None

    # Convert to indices
    indices = [G.idx_for_osm_id(osm_id) for osm_id in path]

    # For each edge: compute bearing, look up name/highway, compute distance
    edges = []
    for i in range(len(indices) - 1):
        u_idx = indices[i]
        v_idx = indices[i + 1]

        lat1 = float(G.node_lats[u_idx])
        lon1 = float(G.node_lons[u_idx])
        lat2 = float(G.node_lats[v_idx])
        lon2 = float(G.node_lons[v_idx])

        b = bearing(lat1, lon1, lat2, lon2)
        d = haversine(lat1, lon1, lat2, lon2)

        name = G.edge_name(u_idx, v_idx)
        highway = G.edge_highway(u_idx, v_idx)

        # Effective name: street name if present, else fallback
        if name:
            effective_name = name
        elif highway:
            effective_name = _HIGHWAY_DESCRIPTIONS.get(highway, "road")
        else:
            effective_name = "road"

        edges.append({
            "bearing": b, "distance": d, "name": name,
            "highway": highway, "effective_name": effective_name,
            "start_idx": i,
        })

    if not edges:
        return None

    # Group consecutive edges with the same effective name
    groups = []
    group_start = 0
    while group_start < len(edges):
        group_end = group_start + 1
        while (group_end < len(edges) and
               edges[group_end]["effective_name"] == edges[group_start]["effective_name"]):
            group_end += 1

        total_dist = sum(edges[j]["distance"] for j in range(group_start, group_end))
        groups.append({
            "effective_name": edges[group_start]["effective_name"],
            "street_name": edges[group_start]["name"],
            "total_distance": total_dist,
            "entry_bearing": edges[group_start]["bearing"],
            "exit_bearing": edges[group_end - 1]["bearing"],
            "start_idx": edges[group_start]["start_idx"],
        })
        group_start = group_end

    # Generate instructions
    steps = []
    for i, group in enumerate(groups):
        node_idx = indices[group["start_idx"]]
        lat = float(G.node_lats[node_idx])
        lon = float(G.node_lons[node_idx])

        if i == 0:
            compass = _compass_direction(group["entry_bearing"])
            instruction = f"Head {compass} on {group['effective_name']}"
            turn_direction = "start"
            turn_angle = 0.0
        else:
            prev_exit = groups[i - 1]["exit_bearing"]
            this_entry = group["entry_bearing"]
            angle = this_entry - prev_exit
            # Normalize to [-180, 180]
            while angle > 180:
                angle -= 360
            while angle < -180:
                angle += 360

            turn_angle = angle
            turn_direction = _classify_turn(angle)
            prefix = _TURN_PREFIXES[turn_direction]
            if turn_direction == "straight":
                instruction = f"{prefix} on {group['effective_name']}"
            else:
                instruction = f"{prefix} onto {group['effective_name']}"

        steps.append({
            "instruction": instruction,
            "street_name": group["street_name"],
            "street_description": group["effective_name"],
            "distance": group["total_distance"],
            "turn_direction": turn_direction,
            "turn_angle": turn_angle,
            "start_lat": lat,
            "start_lon": lon,
        })

    # Final arrive step
    last_idx = indices[-1]
    steps.append({
        "instruction": "Arrive at destination",
        "street_name": None,
        "street_description": "",
        "distance": 0,
        "turn_direction": "arrive",
        "turn_angle": 0,
        "start_lat": float(G.node_lats[last_idx]),
        "start_lon": float(G.node_lons[last_idx]),
    })

    return steps

File: swift-router/Sources/RouterLib/Router.swift
import Foundation

// MARK: - Edge Key

/// Canonical undirected edge key (smaller ID first).
public struct EdgeKey: Hashable {
    public let a: Int64
    public let b: Int64

    public init(_ n1: Int64, _ n2: Int64) {
        if n1 <= n2 {
            a = n1; b = n2
        } else {
            a = n2; b = n1
        }
    }
}

// MARK: - Route Result

public struct RouteResult {
    public let path: [Int64]
    public let edges: [EdgeKey]
    public let distance: Double
    public let shortestDistance: Double
    public let novelty: Double
    public let overhead: Double
    public let instructions: [NavigationStep]?
}

// MARK: - Min-Heap

private struct HeapEntry: Comparable {
    var fScore: Float
    var gScore: Float
    var counter: Int
    var nodeIdx: Int32

    static func < (lhs: HeapEntry, rhs: HeapEntry) -> Bool {
        if lhs.fScore != rhs.fScore { return lhs.fScore < rhs.fScore }
        if lhs.gScore != rhs.gScore { return lhs.gScore < rhs.gScore }
        return lhs.counter < rhs.counter
    }
}

private struct MinHeap {
    private var storage: [HeapEntry] = []

    var isEmpty: Bool { storage.isEmpty }

    mutating func push(_ entry: HeapEntry) {
        storage.append(entry)
        siftUp(storage.count - 1)
    }

    mutating func pop() -> HeapEntry {
        let top = storage[0]
        let last = storage.count - 1
        if last > 0 {
            storage[0] = storage[last]
            storage.removeLast()
            siftDown(0)
        } else {
            storage.removeLast()
        }
        return top
    }

    private mutating func siftUp(_ i: Int) {
        var pos = i
        while pos > 0 {
            let parent = (pos - 1) / 2
            if storage[pos] < storage[parent] {
                storage.swapAt(pos, parent)
                pos = parent
            } else {
                break
            }
        }
    }

    private mutating func siftDown(_ i: Int) {
        var pos = i
        let n = storage.count
        while true {
            var smallest = pos
            let left = 2 * pos + 1
            let right = 2 * pos + 2
            if left < n && storage[left] < storage[smallest] { smallest = left }
            if right < n && storage[right] < storage[smallest] { smallest = right }
            if smallest == pos { break }
            storage.swapAt(pos, smallest)
            pos = smallest
        }
    }
}

// MARK: - A* Heuristic

@inline(__always)
private func heuristicIdx(_ graph: CompactGraph, _ idx: Int32, _ targetLat: Double, _ targetLon: Double) -> Float {
    Float(haversine(lat1: Double(graph.nodeLats[Int(idx)]),
                    lon1: Double(graph.nodeLons[Int(idx)]),
                    lat2: targetLat, lon2: targetLon))
}

// MARK: - Path Reconstruction

private func reconstructPath(_ cameFrom: UnsafeMutablePointer<Int32>,
                              srcIdx: Int32, tgtIdx: Int32) -> [Int32] {
    var path = [Int32]()
    var cur = tgtIdx
    while cur != -1 {
        path.append(cur)
        if cur == srcIdx { break }
        cur = cameFrom[Int(cur)]
    }
    path.reverse()
    return path
}

private func pathDistance(_ graph: CompactGraph, _ pathIndices: [Int32]) -> Double {
    var total = 0.0
    for i in 0..<(pathIndices.count - 1) {
        let u = pathIndices[i]
        let v = pathIndices[i + 1]
        let (targets, weights) = graph.neighbors(u)
        for j in targets.startIndex..<targets.endIndex {
            if targets[j] == v {
                total += Double(weights[j])
                break
            }
        }
    }
    return total
}

private func pathToEdges(_ path: [Int64]) -> [EdgeKey] {
    var edges = [EdgeKey]()
    edges.reserveCapacity(path.count - 1)
    for i in 0..<(path.count - 1) {
        edges.append(EdgeKey(path[i], path[i + 1]))
    }
    return edges
}

private func computeNovelty(_ edges: [EdgeKey], _ walkedEdges: Set<EdgeKey>) -> Double {
    if edges.isEmpty { return 1.0 }
    var novel = 0
    for e in edges {
        if !walkedEdges.contains(e) { novel += 1 }
    }
    return Double(novel) / Double(edges.count)
}

// MARK: - Shortest Path

/// A* shortest path between two OSM node IDs.
/// Returns (path of OSM IDs, distance in meters) or nil.
public func shortestPath(_ graph: CompactGraph, source: Int64, target: Int64)
    -> (path: [Int64], distance: Double)? {

    if source == target { return ([source], 0.0) }

    let srcIdx = graph.idxForOsmId(source)
    let tgtIdx = graph.idxForOsmId(target)
    let targetLat = Double(graph.nodeLats[Int(tgtIdx)])
    let targetLon = Double(graph.nodeLons[Int(tgtIdx)])
    let numNodes = graph.numNodes

    let cameFrom = UnsafeMutablePointer<Int32>.allocate(capacity: numNodes)
    cameFrom.initialize(repeating: -1, count: numNodes)
    defer { cameFrom.deallocate() }

    let gScore = UnsafeMutablePointer<Float>.allocate(capacity: numNodes)
    gScore.initialize(repeating: .infinity, count: numNodes)
    defer { gScore.deallocate() }

    gScore[Int(srcIdx)] = 0.0

    var counter = 0
    var openSet = MinHeap()
    openSet.push(HeapEntry(fScore: 0.0, gScore: 0.0, counter: 0, nodeIdx: srcIdx))

    while !openSet.isEmpty {
        let entry = openSet.pop()
        let current = entry.nodeIdx
        let g = entry.gScore

        if current == tgtIdx {
            let pathIndices = reconstructPath(cameFrom, srcIdx: srcIdx, tgtIdx: tgtIdx)
            let pathOsm = pathIndices.map { graph.nodeIds[Int($0)] }
            cameFrom.deinitialize(count: numNodes)
            gScore.deinitialize(count: numNodes)
            return (pathOsm, Double(g))
        }

        if g > gScore[Int(current)] { continue }

        let (targets, weights) = graph.neighbors(current)
        for j in targets.startIndex..<targets.endIndex {
            let neighbor = targets[j]
            let newG = g + weights[j]

            if newG < gScore[Int(neighbor)] {
                gScore[Int(neighbor)] = newG
                cameFrom[Int(neighbor)] = current
                let h = heuristicIdx(graph, neighbor, targetLat, targetLon)
                counter += 1
                openSet.push(HeapEntry(fScore: newG + h, gScore: newG,
                                       counter: counter, nodeIdx: neighbor))
            }
        }
    }

    cameFrom.deinitialize(count: numNodes)
    gScore.deinitialize(count: numNodes)
    return nil
}

// MARK: - Penalized A*

private func penalizedAstar(_ graph: CompactGraph, source: Int64, target: Int64,
                             walkedEdges: Set<EdgeKey>, penaltyFactor: Float)
    -> (path: [Int64], distance: Double)? {

    if source == target { return ([source], 0.0) }

    let srcIdx = graph.idxForOsmId(source)
    let tgtIdx = graph.idxForOsmId(target)
    let targetLat = Double(graph.nodeLats[Int(tgtIdx)])
    let targetLon = Double(graph.nodeLons[Int(tgtIdx)])
    let numNodes = graph.numNodes

    let cameFrom = UnsafeMutablePointer<Int32>.allocate(capacity: numNodes)
    cameFrom.initialize(repeating: -1, count: numNodes)
    defer { cameFrom.deallocate() }

    let gScore = UnsafeMutablePointer<Float>.allocate(capacity: numNodes)
    gScore.initialize(repeating: .infinity, count: numNodes)
    defer { gScore.deallocate() }

    gScore[Int(srcIdx)] = 0.0

    var counter = 0
    var openSet = MinHeap()
    openSet.push(HeapEntry(fScore: 0.0, gScore: 0.0, counter: 0, nodeIdx: srcIdx))

    while !openSet.isEmpty {
        let entry = openSet.pop()
        let current = entry.nodeIdx
        let g = entry.gScore

        if current == tgtIdx {
            let pathIndices = reconstructPath(cameFrom, srcIdx: srcIdx, tgtIdx: tgtIdx)
            let pathOsm = pathIndices.map { graph.nodeIds[Int($0)] }
            let actualDist = pathDistance(graph, pathIndices)
            cameFrom.deinitialize(count: numNodes)
            gScore.deinitialize(count: numNodes)
            return (pathOsm, actualDist)
        }

        if g > gScore[Int(current)] { continue }

        let currentOsm = graph.nodeIds[Int(current)]
        let (targets, weights) = graph.neighbors(current)
        for j in targets.startIndex..<targets.endIndex {
            let neighbor = targets[j]
            let edgeWeight = weights[j]

            let neighborOsm = graph.nodeIds[Int(neighbor)]
            let ek = EdgeKey(currentOsm, neighborOsm)

            var effectiveWeight = edgeWeight
            if walkedEdges.contains(ek) {
                effectiveWeight *= penaltyFactor
            }

            let newG = g + effectiveWeight
            if newG < gScore[Int(neighbor)] {
                gScore[Int(neighbor)] = newG
                cameFrom[Int(neighbor)] = current
                let h = heuristicIdx(graph, neighbor, targetLat, targetLon)
                counter += 1
                openSet.push(HeapEntry(fScore: newG + h, gScore: newG,
                                       counter: counter, nodeIdx: neighbor))
            }
        }
    }

    cameFrom.deinitialize(count: numNodes)
    gScore.deinitialize(count: numNodes)
    return nil
}

// MARK: - Novelty Route

private func buildResult(path: [Int64], distance: Double, baseDistance: Double,
                          walkedEdges: Set<EdgeKey>, graph: CompactGraph? = nil) -> RouteResult {
    let edges = pathToEdges(path)
    let novelty = computeNovelty(edges, walkedEdges)
    let overhead = baseDistance > 0 ? (distance - baseDistance) / baseDistance : 0
    let instructions: [NavigationStep]?
    if let g = graph, g.hasNameData {
        instructions = generateInstructions(path: path, graph: g)
    } else {
        instructions = nil
    }
    return RouteResult(path: path, edges: edges, distance: distance,
                       shortestDistance: baseDistance, novelty: novelty,
                       overhead: overhead, instructions: instructions)
}

/// Find a route maximizing novel (unwalked) edges.
public func noveltyRoute(_ graph: CompactGraph, source: Int64, target: Int64,
                          walkedEdges: Set<EdgeKey>,
                          minNovelty: Double = 0.3, maxOverhead: Double = 0.25)
    -> RouteResult? {

    // Phase 1: baseline shortest path
    guard let base = shortestPath(graph, source: source, target: target) else {
        return nil
    }

    let baseEdges = pathToEdges(base.path)
    let baseNovel = computeNovelty(baseEdges, walkedEdges)

    // Short-circuit only if baseline meets novelty AND the budget is small.
    // When the user asks for significant extra time (maxOverhead > 0.3),
    // proceed to the via-waypoint phase to produce a longer route.
    if baseNovel >= minNovelty && maxOverhead < 0.3 {
        return buildResult(path: base.path, distance: base.distance,
                           baseDistance: base.distance, walkedEdges: walkedEdges, graph: graph)
    }

    // Phase 2: iterative penalty search — find upper bound where novelty is met
    var bestResult: RouteResult? = buildResult(path: base.path, distance: base.distance,
                                               baseDistance: base.distance, walkedEdges: walkedEdges, graph: graph)
    var bestOverhead: Double = 0
    var bestMeetsNovelty = baseNovel >= minNovelty

    var loPenalty: Float = 1.0
    var hiPenalty: Float = 10.0

    // Skip penalty-based search when there are no walked edges (penalty has no effect)
    if walkedEdges.isEmpty {
        // Jump straight to Phase 5 (via-waypoint lengthening)
    } else {

    for _ in 0..<5 {
        guard let r = penalizedAstar(graph, source: source, target: target,
                                     walkedEdges: walkedEdges, penaltyFactor: hiPenalty) else {
            hiPenalty = (loPenalty + hiPenalty) / 2
            continue
        }
        let edges = pathToEdges(r.path)
        let novelty = computeNovelty(edges, walkedEdges)
        if novelty >= minNovelty { break }
        hiPenalty *= 2
        if hiPenalty > 100 { break }
    }

    // Phase 3: binary search — among routes meeting novelty, prefer those
    // that use the most of the time budget (highest overhead within limit)
    for _ in 0..<10 {
        let midPenalty = (loPenalty + hiPenalty) / 2
        guard let r = penalizedAstar(graph, source: source, target: target,
                                     walkedEdges: walkedEdges, penaltyFactor: midPenalty) else {
            hiPenalty = midPenalty
            continue
        }

        let edges = pathToEdges(r.path)
        let novelty = computeNovelty(edges, walkedEdges)
        let overhead = base.distance > 0 ? (r.distance - base.distance) / base.distance : 0
        let meetsNovelty = novelty >= minNovelty

        // Keep this result if it's better than what we have:
        // 1. First result meeting novelty beats any that don't
        // 2. Among results meeting novelty, prefer highest overhead (use the budget)
        // 3. Among results not meeting novelty, prefer highest novelty
        let dominated: Bool
        if meetsNovelty && overhead <= maxOverhead {
            dominated = bestMeetsNovelty && overhead <= bestOverhead
        } else if !meetsNovelty && overhead <= maxOverhead {
            dominated = bestMeetsNovelty || overhead <= bestOverhead
        } else {
            dominated = true // over budget
        }

        if !dominated {
            bestOverhead = overhead
            bestMeetsNovelty = meetsNovelty
            bestResult = buildResult(path: r.path, distance: r.distance,
                                     baseDistance: base.distance, walkedEdges: walkedEdges, graph: graph)
        }

        if novelty < minNovelty {
            loPenalty = midPenalty
        } else if overhead > maxOverhead {
            hiPenalty = midPenalty
        } else {
            loPenalty = midPenalty
        }
    }

    // Phase 4: fallback fixed penalties
    if bestResult == nil || !bestMeetsNovelty {
        for penalty: Float in [1.5, 2.0, 3.0, 5.0, 8.0] {
            guard let r = penalizedAstar(graph, source: source, target: target,
                                         walkedEdges: walkedEdges, penaltyFactor: penalty) else {
                continue
            }
            let edges = pathToEdges(r.path)
            let novelty = computeNovelty(edges, walkedEdges)
            let overhead = base.distance > 0 ? (r.distance - base.distance) / base.distance : 0
            let meetsNovelty = novelty >= minNovelty

            if overhead <= maxOverhead {
                let dominated: Bool
                if meetsNovelty {
                    dominated = bestMeetsNovelty && overhead <= bestOverhead
                } else {
                    dominated = bestMeetsNovelty || overhead <= bestOverhead
                }
                if !dominated {
                    bestOverhead = overhead
                    bestMeetsNovelty = meetsNovelty
                    bestResult = buildResult(path: r.path, distance: r.distance,
                                             baseDistance: base.distance, walkedEdges: walkedEdges)
                }
            }
        }
    }

    } // end if !walkedEdges.isEmpty

    // Phase 5: via-waypoint lengthening
    // If the best route is far shorter than the budget allows, route via an
    // intermediate waypoint offset perpendicular to the start→end line.
    // Try multiple offset scales since road distance >> straight-line distance.
    let bestDist = bestResult?.distance ?? base.distance
    let targetDist = base.distance * (1 + maxOverhead)

    if bestDist < targetDist * 0.85 {
        let srcIdx = graph.idxForOsmId(source)
        let tgtIdx = graph.idxForOsmId(target)
        let sLat = Double(graph.nodeLats[Int(srcIdx)])
        let sLon = Double(graph.nodeLons[Int(srcIdx)])
        let eLat = Double(graph.nodeLats[Int(tgtIdx)])
        let eLon = Double(graph.nodeLons[Int(tgtIdx)])

        let midLat = (sLat + eLat) / 2
        let midLon = (sLon + eLon) / 2
        let cosLat = cos(midLat * .pi / 180)

        // Direction vector (equirectangular)
        let dx = (eLon - sLon) * cosLat
        let dy = eLat - sLat
        let dirLen = sqrt(dx * dx + dy * dy)

        if dirLen > 0 {
            // Perpendicular unit vector
            let px = -dy / dirLen
            let py = dx / dirLen

            // Ideal straight-line offset from triangle geometry.
            // Road routing adds ~40% over straight-line, so we try several
            // scales from small to large to find one that fits the budget.
            let d = base.distance
            let D = targetDist
            let hIdeal = D > d ? sqrt(D * D - d * d) / 2 : d * 0.3

            let scales: [Double] = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.15]
            let signs: [Double] = [1.0, -1.0]

            print("[Route] Via-waypoint: baseline=\(Int(d))m target=\(Int(D))m hIdeal=\(Int(hIdeal))m bestSoFar=\(Int(bestDist))m")

            for scale in scales {
                let h = hIdeal * scale
                let hDeg = h / 111320.0

                for sign in signs {
                    let wpLat = midLat + sign * px * hDeg
                    let wpLon = midLon + sign * py * hDeg / cosLat

                    let wpSnap = graph.findNearestNode(lat: wpLat, lon: wpLon)
                    let wpId = graph.nodeIds[Int(wpSnap.index)]

                    if wpId == source || wpId == target { continue }

                    guard let leg1 = shortestPath(graph, source: source, target: wpId),
                          let leg2 = shortestPath(graph, source: wpId, target: target)
                    else { continue }

                    let combinedDist = leg1.distance + leg2.distance
                    let combinedOverhead = d > 0 ? (combinedDist - d) / d : 0

                    print("[Route]   scale=\(scale) sign=\(Int(sign)) dist=\(Int(combinedDist))m overhead=\(String(format: "%.0f", combinedOverhead * 100))%")

                    if combinedOverhead <= maxOverhead && combinedOverhead > bestOverhead {
                        let combinedPath = leg1.path + leg2.path.dropFirst()
                        bestResult = buildResult(path: combinedPath, distance: combinedDist,
                                                 baseDistance: d, walkedEdges: walkedEdges,
                                                 graph: graph)
                        bestOverhead = combinedOverhead
                    }
                }
            }

            print("[Route] Via-waypoint result: \(bestOverhead > 0 ? "\(Int(bestResult?.distance ?? 0))m" : "none fit budget")")
        }
    }

    // Phase 6: worst-case fallback
    if bestResult == nil {
        bestResult = buildResult(path: base.path, distance: base.distance,
                                 baseDistance: base.distance, walkedEdges: walkedEdges, graph: graph)
    }

    return bestResult
}
